Master tracking payload requirements

rvykydal opened the pull request.

rvykydal commented:
jenkins, test this please

M4rtinK commented on pyanaconda/payload/livepayload.py:
```diff
@@ -98,9 +98,9 @@ def unsetup(self):
         except OSError:
             pass
 
```
I think it would make sense to log a warning if there are packages/groups requested to be added but the current payload is not package based/does not support adding additional packages.

I'm not sure where the best place for this is though - I guess either when attempts are made to add the packages/groups or when the payload API methods that would normally install the packages/groups are executed.

rvykydal commented:
PR updated with these changes based on the review
The first patch (1/5):
- types of requirements defined by Enum
- _add function moved for better readability
- log message with reason for not selecting a package ("IGNORED") can contain more reasons
- updated documentation of requirement attribute "strong"
- __str__ method improved
The patches 2/5-4/5 stay the same.
Added patch 5/5 for tracking application of requests, hopefully also addressing Jirka's remark on SRP principle.

rvykydal commented:
Packaging log for dnf payload with -lvm2 in %packages:
https://rvykydal.fedorapeople.org/packaging.reqs.log
If you grep it for "requirement" you get:
```
10:09:46,780 DEBUG packaging: added package requirement 'lvm2' for storage, strong=True
10:09:46,783 DEBUG packaging: added package requirement 'xfsprogs' for storage, strong=True
10:09:46,783 DEBUG packaging: added package requirement 'e2fsprogs' for storage, strong=True
10:09:46,783 DEBUG packaging: no package requirement added for realm
10:09:46,784 DEBUG packaging: no package requirement added for authconfig
10:09:46,784 DEBUG packaging: no package requirement added for firewall
10:09:46,785 DEBUG packaging: no package requirement added for network
10:09:46,785 DEBUG packaging: no package requirement added for ntp
10:09:46,786 DEBUG packaging: added package requirement 'grub2' for bootloader, strong=True
10:09:46,787 DEBUG packaging: no group requirement added for language groups
10:09:46,795 DEBUG packaging: no package requirement added for langpacks
10:09:48,211 DEBUG packaging: selected package: lvm2, requirement for ['storage'] IGNORED because excluded
10:09:48,219 DEBUG packaging: selected package: xfsprogs, requirement for ['storage'] 
10:09:48,227 DEBUG packaging: selected package: e2fsprogs, requirement for ['storage'] 
10:09:48,235 DEBUG packaging: selected package: grub2, requirement for ['bootloader'] 
10:09:48,236 DEBUG packaging: apply with result True called on requirements [('package', 'lvm2', PayloadRequirement(id='lvm2', reasons=['storage'], strong=True)), ('package', 'xfsprogs', PayloadRequirement(id='xfsprogs', reasons=['storage'], strong=True)), ('package', 'e2fsprogs', PayloadRequirement(id='e2fsprogs', reasons=['storage'], strong=True)), ('package', 'grub2', PayloadRequirement(id='grub2', reasons=['bootloader'], strong=True))]
10:17:17,084 INFO packaging: Installation requirements: [('package', 'lvm2', PayloadRequirement(id='lvm2', reasons=['storage'], strong=True)), ('package', 'xfsprogs', PayloadRequirement(id='xfsprogs', reasons=['storage'], strong=True)), ('package', 'e2fsprogs', PayloadRequirement(id='e2fsprogs', reasons=['storage'], strong=True)), ('package', 'grub2', PayloadRequirement(id='grub2', reasons=['bootloader'], strong=True))]
```
For payload which doesn't call requirements.apply() you get:
```
10:09:46,780 DEBUG packaging: added package requirement 'lvm2' for storage, strong=True
10:09:46,783 DEBUG packaging: added package requirement 'xfsprogs' for storage, strong=True
10:09:46,783 DEBUG packaging: added package requirement 'e2fsprogs' for storage, strong=True
10:09:46,783 DEBUG packaging: no package requirement added for realm
10:09:46,784 DEBUG packaging: no package requirement added for authconfig
10:09:46,784 DEBUG packaging: no package requirement added for firewall
10:09:46,785 DEBUG packaging: no package requirement added for network
10:09:46,785 DEBUG packaging: no package requirement added for ntp
10:09:46,786 DEBUG packaging: added package requirement 'grub2' for bootloader, strong=True
10:09:46,787 DEBUG packaging: no group requirement added for language groups
10:09:46,795 DEBUG packaging: no package requirement added for langpacks
10:09:48,211 DEBUG packaging: selected package: lvm2, requirement for ['storage'] IGNORED because excluded
10:09:48,219 DEBUG packaging: selected package: xfsprogs, requirement for ['storage'].
10:09:48,227 DEBUG packaging: selected package: e2fsprogs, requirement for ['storage'].
10:09:48,235 DEBUG packaging: selected package: grub2, requirement for ['bootloader'].
10:17:17,084 INFO packaging: Installation requirements: [('package', 'lvm2', PayloadRequirement(id='lvm2', reasons=['storage'], strong=True)), ('package', 'xfsprogs', PayloadRequirement(id='xfsprogs', reasons=['storage'], strong=True)), ('package', 'e2fsprogs', PayloadRequirement(id='e2fsprogs', reasons=['storage'], strong=True)), ('package', 'grub2', PayloadRequirement(id='grub2', reasons=['bootloader'], strong=True))]
10:17:17,085 INFO packaging: Some of the requirements were not applied.
```

jkonecny12 commented on pyanaconda/payload/__init__.py:
```diff
@@ -137,6 +139,89 @@ class DependencyError(PayloadError):
+        :type group_ids: list of str
+        :param reason: descripiton of reason for adding the groups
+        :type reason: str
+        :param strong: is the requirement strong
```
Could you please add the `` (ie is not satisfying it fatal?)`` message here too.

jkonecny12 commented:
Please fix the tests before merging this.

jkonecny12 commented:
Do you think @rvykydal that the https://github.com/rhinstaller/anaconda/pull/917 could use of the changes here?

rvykydal commented:
@jkonecny12 I don't think so, it is a package exclusion required by user, not by installation process/functionality itself.

rvykydal commented:
jenkins, test this please

rvykydal commented:
jenkins, test this please

rvykydal commented:
jenkins, test this please

rvykydal commented:
jenkins, test this please

rvykydal commented:
jenkins, test this please

rvykydal commented:
So finally the checks have passed. To achieve that I had to use a class instead of namedtuple for PayloadRequirement as it is actually not an immutable object, ie reasons for the requirement can be added later on after creating the requirement (which was possible using namedtuple because the reasons were stored in a list item of the tuple) but also strength of the requirement can be updated and this couldn't be addressed using bool tuple item.
As a little bonus, strengths of individual reasons for requirement (not only the resulting strength of the whole requirement as before) are now tracked in the PayloadRequirement object which can be handy.

The change worth review concerns only the first patch - replacing PayloadRequirement namedtuple with a class. Here is the diff:
https://rvykydal.fedorapeople.org/payload_requirement.patch

jkonecny12 commented on pyanaconda/payload/__init__.py:
```diff
@@ -137,6 +139,179 @@ class DependencyError(PayloadError):
+    pass
+
+class PayloadRequirements(object):
+    """A container for payload requirements imposed by installed functionality
```
Missing a `.` period here.

rvykydal commented:
jenkins, test this please
